### PR TITLE
Add golang sample code

### PR DESCRIPTION
**Pull request by MSGaryWang**
For authentication sample code piece

### CODE REVIEW

File: docs/REST/authentication.md [hunk 1]
```diff
@@ -348,4 +348,47 @@ private static String buildContentHash(HttpUriRequest request) throws IOExceptio
...
+}
+
+func signRequest(content string, key []byte) string {
+    hmac := hmac.New(sha256.New, key)
+    hmac.Write([]byte(content))
+    return base64.StdEncoding.EncodeToString(hmac.Sum(nil))
+}
+
+// Setup the auth header for accessing Azure AppConfiguration service
+func PrepareAuthHeader(verb string, host string, pathAndQuery string, body string, id string, secret string, req *http.Request) {
```
Comment (jimmyca15) [inline docs/REST/authentication.md]: >verb string, host string, pathAndQuery string, body string [](start = 23, length = 58)

Is it common to pass around everything when working with HTTP request in go?

I would expect just passing in (req, id, secret) like in java or C#. Then getting req.Method, req.Host, req.URL, and req.GetBody() to prepare the authentication. With all the current parameters it kind of seems like the httprequest should be generated in the method rather than passed in.

File: docs/REST/authentication.md [hunk 2]
```diff
@@ -348,4 +348,47 @@ private static String buildContentHash(HttpUriRequest request) throws IOExceptio
...
+
+    timestamp := time.Now().UTC().Format(http.TimeFormat)
+    contentHash := getContentHashBase64(body)
+    signature := signRequest(fmt.Sprintf("%s\n%s\n%s;%s;%s", strings.ToUpper(verb), pathAndQuery, timestamp, host, contentHash), key)
+
+    req.Header.Set("x-ms-content-sha256", contentHash)
+    req.Header.Set("x-ms-date", timestamp)
+    req.Header.Set("Authorization", "HMAC-SHA256 Credential="+id+", SignedHeaders=x-ms-date;host;x-ms-content-sha256, Signature="+signature)
+    req.Header.Set("Content-Type", "application/vnd.microsoft.appconfig.kv+json")
+}
```
Comment (jimmyca15) [inline docs/REST/authentication.md]: This is not related to authentication.

File: docs/REST/authentication.md [hunk 3]
```diff
@@ -348,4 +348,47 @@ private static String buildContentHash(HttpUriRequest request) throws IOExceptio
...
+// Setup the auth header for accessing Azure AppConfiguration service
+func PrepareAuthHeader(verb string, host string, pathAndQuery string, body string, id string, secret string, req *http.Request) {
+    key, err := base64.StdEncoding.DecodeString(secret)
+    if err != nil {
+        fmt.Printf("err = %s \n", err)
+    }
+
+    timestamp := time.Now().UTC().Format(http.TimeFormat)
+    contentHash := getContentHashBase64(body)
+    signature := signRequest(fmt.Sprintf("%s\n%s\n%s;%s;%s", strings.ToUpper(verb), pathAndQuery, timestamp, host, contentHash), key)
```
Comment (zhenlan) [inline docs/REST/authentication.md]: Just like other examples, it's better to split this to separate lines to have a `stringToSign`, which will be easier to read.

File: docs/REST/authentication.md [hunk 4]
```diff
@@ -348,4 +348,47 @@ private static String buildContentHash(HttpUriRequest request) throws IOExceptio
...
+    "time"
+)
+
+func getContentHashBase64(content string) string {
+    hasher := sha256.New()
+    hasher.Write([]byte(content))
+    return base64.StdEncoding.EncodeToString(hasher.Sum(nil))
+}
+
+func signRequest(content string, key []byte) string {
```
Comment (zhenlan) [inline docs/REST/authentication.md]: Technically, this is not signing a request. It's just calculating hmac of a given string. Maybe `getHmac` is better?

Can these helper functions be put to the end of the code or do they have to be declared before called in go?

File: docs/REST/authentication.md (hunk 1 above)
Comment (zhenlan) [reply docs/REST/authentication.md]: BTW, it will be easier to read to just call this function `Sign` or `SignRequest`. It will be consistent with other examples.

Comment (zhenlan) [reply docs/REST/authentication.md]: Do you mind to change the `execute` in Java example to `SignRequest` too?
